### PR TITLE
Open marker popups when navigating with the keyboard

### LINKED ISSUE #5071: Accessibility Issue: Map doesn't work fully when navigating with keyboard
Background:

One of our councils has been performing some Accessibility tests and noticed that when you're navigating an investment page using the tab button, you can move from one pin to another but it's not possible to select a pin.

Example:
https://dundeesvoice.communitychoices.scot/budgets/7

image

We're not sure if this is a limitation of OpenStreetMaps or the implementation but it would be good if this map was fully functional when navigating using the keyboard.

### PR DESCRIPTION
**Pull request by Senen**
## References

* Closes #5071
* Depends on #5108

## Objectives
Allow opening marker popups when navigating with the keyboard.

## How
Use the `bindPopup` Leaflet method as the documentation says as it handles the keyboard event properly.

OpenMarkerPopupWithKeyboard

### CODE REVIEW

**Comment by javierm** [comment]
@Senen I've taken the liberty to rebase this branch against the `map_refactoring` branch (PR #5108), since we were both touching code affecting the `getPopupContent` function.

**Review by javierm — COMMENTED** [review]
I'm very sorry :pray:. I wrote this review two months ago and thought I had submitted it, but now I see I moved the PR to the "Doing" column but left the review as a draft :sweat_smile:. Sorry!

**Comment by Senen** [comment]
Hi Javier, I tried a different approach that binds the popup on marker loading but waits for the user interaction to load its content.

**Review by javierm — COMMENTED** [review]
It looks like the popup behaves differently now and there isn't enough width for the content of the popup:

The marker popup displays one word per line

I think this is because we bind the popup with empty content and so its width is set to the minimum, and when we populate the content, its width doesn't change.

The other difference I've noticed is that now the popup is displayed with blank content while we're performing the AJAX call :thinking:.

Two possible solutions come to mind.

First, we could perform a "syncrhonous asynchonous call" :sweat_smile: to get the content. Here's the patch.


```diff
diff --git a/app/assets/javascripts/map.js b/app/assets/javascripts/map.js
index 8fa3ecc713..4fcbd2edd7 100644
--- a/app/assets/javascripts/map.js
+++ b/app/assets/javascripts/map.js
@@ -173,9 +173,7 @@
             marker.options.id = coordinates.investment_id;
 
             marker.bindPopup(function() {
-              var popupContent = document.createElement("div");
-              App.Map.openMarkerPopup(popupContent, marker.options.id);
-              return popupContent;
+              return App.Map.openMarkerPopup(marker.options.id);
             });
           }
         });
@@ -199,14 +197,19 @@
       map.attributionControl.setPrefix(App.Map.attributionPrefix());
       L.tileLayer(mapTilesProvider, { attribution: mapAttribution }).addTo(map);
     },
-    openMarkerPopup: function(content, id) {
+    openMarkerPopup: function(id) {
+      var content;
+
       $.ajax("/investments/" + id + "/json_data", {
         type: "GET",
+        async: false,
         dataType: "json",
         success: function(data) {
-          content.innerHTML = App.Map.getPopupContent(data);
+          content = App.Map.getPopupContent(data);
         }
       });
+
+      return content;
     },
     getPopupContent: function(data) {
       return "<a href='/budgets/" + data.budget_id + "/investments/" + data.investment_id + "'>" +
```

A possible problem with this implementation is that the browser cannot do anything else while doing the synchronous call, which might be an issue if the connection is slow :thinking:.

The second option would be to keep doing what we were doing (binding the popup on the "click" event") but doing so when the "Enter" key is pressed as well:

```javascript
marker.on("keydown", function(e) {
  if (e.originalEvent.keyCode === 13) {
    App.Map.openMarkerPopup(e);
  }
});
```

No idea why `e.keyCode` doesn't work and I had to add `originalEvent`.

**Comment by javierm** [comment]
Just for the record, here's a method to get the JSON data that seems to perform as in the `master` branch and preloads the content of the popup so we can use `bindPopup` and forget about AJAX calls.

The downside is that the code is complex in order to deal with the internationalization fallbacks for the title :relieved:.

```ruby
  def self.investments_json_data(investments)
    return [] unless investments.any?

    budget_id = investments.first.budget_id
    locales_with_order = Globalize.fallbacks(I18n.locale).map.with_index  { |locale, order| "('#{locale}', #{order})" }.join(", ")

    data = investments
            .joins(:map_location)
            .with_translations(Globalize.fallbacks(I18n.locale))
            .joins("LEFT JOIN (VALUES #{locales_with_order}) AS locales(locale, ordering) ON #{Budget::Investment.translations_table_name}.locale = locales.locale")
            .order("locales.ordering")
            .pluck(:id, :title, :latitude, :longitude, "locales.ordering")

    data.uniq { |values| values[0] }.map do |values|
      {
        title: values[1],
        link: "/budgets/#{budget_id}/investments/#{values[0]}",
        lat: values[2],
        long: values[3]
      }
    end
  end
```

Funny enough, one of the bottlenecks turned out to be calling the `polymorphic_path` (or `budget_investment_path`) method thousands  of times :open_mouth:, so that's why the link is generated in the same manual way we generate it in JavaScript.

**Comment by javierm** [comment]
Here's a  version which extracts the fallback logic to the `Globalizable` concern:

```diff
diff --git a/app/components/budgets/map_component.rb b/app/components/budgets/map_component.rb
index 1da16c1378..9e5e686ef0 100644
--- a/app/components/budgets/map_component.rb
+++ b/app/components/budgets/map_component.rb
@@ -21,6 +21,6 @@ class Budgets::MapComponent < ApplicationComponent
         investments = budget.investments
       end
 
-      MapLocation.where(investment_id: investments).map(&:json_data)
+      MapLocation.investments_json_data(investments)
     end
 end
diff --git a/app/controllers/budgets/investments_controller.rb b/app/controllers/budgets/investments_controller.rb
index 3d772820e0..dc5aeeeba6 100644
--- a/app/controllers/budgets/investments_controller.rb
+++ b/app/controllers/budgets/investments_controller.rb
@@ -40,7 +40,7 @@ module Budgets
       @investments = investments.page(params[:page]).per(PER_PAGE).for_render
 
       @investment_ids = @investments.ids
-      @investments_map_coordinates = MapLocation.where(investment: investments).map(&:json_data)
+      @investments_map_coordinates = MapLocation.investments_json_data(investments)
 
       @tag_cloud = tag_cloud
       @remote_translations = detect_remote_translations(@investments)
diff --git a/app/models/concerns/globalizable.rb b/app/models/concerns/globalizable.rb
index c2e04554f1..e33e347ed8 100644
--- a/app/models/concerns/globalizable.rb
+++ b/app/models/concerns/globalizable.rb
@@ -100,5 +100,22 @@ module Globalizable
     def translation_class_delegate(method)
       translation_class.instance_eval { delegate method, to: :globalized_model }
     end
+
+    def with_fallback_translation
+      translations_foreign_key = reflect_on_association(:translations).foreign_key
+      fallbacks = Globalize.fallbacks(Globalize.locale)
+
+      fallbacks_with_order = fallbacks.map.with_index  do |locale, order|
+        "('#{locale}', #{order})"
+      end.join(", ")
+
+      translations_ids = translation_class
+        .select("DISTINCT ON (#{translations_foreign_key}) id")
+        .where(locale: fallbacks)
+        .joins("LEFT JOIN (VALUES #{fallbacks_with_order}) AS locales(name, ordering) ON #{translations_table_name}.locale = locales.name")
+        .order(translations_foreign_key, "locales.ordering")
+
+      with_translations(fallbacks).where("#{translations_table_name}.id": translations_ids)
+    end
   end
 end
diff --git a/app/models/map_location.rb b/app/models/map_location.rb
index 3b441d954c..1fdd8c90ee 100644
--- a/app/models/map_location.rb
+++ b/app/models/map_location.rb
@@ -17,6 +17,25 @@ class MapLocation < ApplicationRecord
     }
   end
 
+  def self.investments_json_data(investments)
+    return [] unless investments.any?
+
+    investments ||= Budget.last.investments;
+
+    budget_id = investments.first.budget_id
+
+    data = investments.joins(:map_location).with_fallback_translation.pluck(:id, :title, :latitude, :longitude)
+
+    data.map do |values|
+      {
+        title: values[1],
+        link: "/budgets/#{budget_id}/investments/#{values[0]}",
+        lat: values[2],
+        long: values[3]
+      }
+    end
+  end
+
   def self.from_heading(heading)
     new(
       zoom: Budget::Heading::OSM_DISTRICT_LEVEL_ZOOM,
```

File: app/models/concerns/globalizable.rb [hunk 1]
```diff
@@ -100,5 +100,22 @@ def validates_translation(method, options = {})
...
+      fallbacks = Globalize.fallbacks(Globalize.locale)
+
+      fallbacks_with_order = fallbacks.map.with_index do |locale, order|
+        "('#{locale}', #{order})"
+      end.join(", ")
+
+      translations_ids = translation_class
+        .select("DISTINCT ON (#{translations_foreign_key}) id")
+        .where(locale: fallbacks)
+        .joins("LEFT JOIN (VALUES #{fallbacks_with_order}) AS locales(name, ordering) ON locale = locales.name")
```
[the remark this answers is not in the document]
Comment (Senen) [reply app/models/concerns/globalizable.rb:115]: I can live with it! 😉

**Comment by Senen** [comment]
Hi @javierm, I'm finally using your proposed method for the `globalize_models`. It seems to work fine. Let me know what you think.

File: app/components/budgets/investments/map_component.rb [hunk 2]
```diff
@@ -18,7 +18,7 @@ def map_location
     end
 
     def coordinates
-      MapLocation.where(investment: investments).map(&:json_data)
+      MapLocation.investments_json_data(investments.unscope(:order))
```
[the remark this answers is not in the document]
Comment (Senen) [reply app/components/budgets/investments/map_component.rb:21]: The investment collection for the budget investments page is calculated [here](https://github.com/consul/consul/blob/master/app/controllers/budgets/investments_controller.rb#L165), and it adds some ordering clauses that do not work with the new `with_fallback_translation` query.

Maybe a better place where we can do the same is [here](https://github.com/consul/consul/blob/master/app/controllers/budgets/investments_controller.rb#L145).

File: spec/models/concerns/globalizable.rb [hunk 3]
```diff
@@ -188,4 +185,78 @@
...
+      begin
+        I18n.enforce_available_locales = false
+        add_translation(record, :es, "Contenido en español")
+        add_translation(record, :ca, "Contingut en català")
+        add_translation(record, :val, "Contingut en valencià")
+
+        I18n.with_locale(:es) do
+          matching_record = fetch_records_and_return_the_one_with_given_id(record)
+
+          expect(matching_record.send(attribute)).to eq("Contenido en español")
```
[the remark this answers is not in the document]
Comment (Senen) [reply spec/models/concerns/globalizable.rb]: 😅

**Review by javierm — APPROVED** [review]

**Pull request merged**